File: buque/core/tree_builder.py
from __future__ import annotations

from dataclasses import dataclass
import re

from buque.core.models import CandidateHeading, TocNode
from buque.core.scorer import normalize_title

_STRUCTURAL_HEADING_RE = re.compile(r"^\s*(第[零一二三四五六七八九十百千万〇\d]+[卷章])")


@dataclass(slots=True)
class BuildTreeResult:
    toc_nodes: list[TocNode]
    rejected_nodes: list[dict[str, object]]


def build_toc_nodes(candidates: list[CandidateHeading], *, max_level_jump: int = 1) -> BuildTreeResult:
    sorted_candidates = sorted(candidates, key=lambda item: (item.page_index, item.bbox[1], item.bbox[0]))
    seen: set[tuple[int, str]] = set()
    toc_nodes: list[TocNode] = []
    rejected: list[dict[str, object]] = []
    seen_structural_keys: set[str] = set()
    prev_level: int | None = None
    for candidate in sorted_candidates:
        title = normalize_title(candidate.text)
        if not title:
            rejected.append(_reject(candidate, "empty_title"))
            continue

        dedupe_key = (candidate.page_index, title.lower())
        if dedupe_key in seen:
            rejected.append(_reject(candidate, "duplicate_same_page"))
            continue
        seen.add(dedupe_key)

        structural_key = _structural_key(title)
        if structural_key and structural_key in seen_structural_keys:
            rejected.append(_reject(candidate, "duplicate_structural_heading"))
            continue
        if structural_key:
            seen_structural_keys.add(structural_key)

        level = max(1, min(6, candidate.level_hint or 1))
        if prev_level is None and level > 1:
            level = 1
        if prev_level is not None and level > prev_level + max_level_jump:
            level = prev_level + max_level_jump
        prev_level = level

        toc_nodes.append(
            TocNode(
                title=title,
                level=level,
                page_index=max(0, candidate.page_index),
                confidence=max(0.0, min(1.0, candidate.total_score)),
                source="rule",
            )
        )

    return BuildTreeResult(toc_nodes=toc_nodes, rejected_nodes=rejected)


def _reject(candidate: CandidateHeading, reason: str) -> dict[str, object]:
    return {
        "page_index": candidate.page_index,
        "title": candidate.text,
        "reason": reason,
        "score": candidate.total_score,
    }


def _structural_key(title: str) -> str:
    matched = _STRUCTURAL_HEADING_RE.match(title.replace(" ", ""))
    if not matched:
        return ""
    return matched.group(1)

File: src/core/tree-builder.ts
import type { CandidateHeading, TocNode } from "../types.js";
import { normalizeTitle } from "./scorer.js";

const STRUCTURAL_HEADING_RE = /^\s*(第[零一二三四五六七八九十百千万〇\d]+[卷章])/u;
const CHINESE_VOLUME_RE = /^第[零一二三四五六七八九十百千万〇\d]+卷/u;
const CHINESE_CHAPTER_RE = /^第[零一二三四五六七八九十百千万〇\d]+章/u;

export interface BuildTreeResult {
  tocNodes: TocNode[];
  rejectedNodes: Array<Record<string, unknown>>;
}

export function buildTocNodes(candidates: CandidateHeading[], options: { maxLevelJump?: number } = {}): BuildTreeResult {
  const maxLevelJump = options.maxLevelJump ?? 1;
  const sortedCandidates = [...candidates].sort((left, right) => {
    return left.pageIndex - right.pageIndex || left.bbox[1] - right.bbox[1] || left.bbox[0] - right.bbox[0];
  });
  const seen = new Set<string>();
  const seenStructuralKeys = new Set<string>();
  const tocNodes: TocNode[] = [];
  const rejectedNodes: Array<Record<string, unknown>> = [];
  let previousLevel: number | undefined;
  let seenVolume = false;

  for (const candidate of sortedCandidates) {
    const title = normalizeTitle(candidate.text);
    if (!title) {
      rejectedNodes.push(reject(candidate, "empty_title"));
      continue;
    }

    const dedupeKey = `${candidate.pageIndex}\u0000${title.toLowerCase()}`;
    if (seen.has(dedupeKey)) {
      rejectedNodes.push(reject(candidate, "duplicate_same_page"));
      continue;
    }
    seen.add(dedupeKey);

    const structuralKey = structuralKeyFor(title);
    if (structuralKey && seenStructuralKeys.has(structuralKey)) {
      rejectedNodes.push(reject(candidate, "duplicate_structural_heading"));
      continue;
    }
    if (structuralKey) seenStructuralKeys.add(structuralKey);

    const compact = title.replace(/\s+/gu, "");
    let level = Math.max(1, Math.min(6, candidate.levelHint ?? 1));
    if (CHINESE_VOLUME_RE.test(compact)) {
      seenVolume = true;
      level = 1;
    } else if (seenVolume && CHINESE_CHAPTER_RE.test(compact)) {
      level = Math.max(level, 2);
    }
    if (previousLevel === undefined && level > 1) level = 1;
    if (previousLevel !== undefined && level > previousLevel + maxLevelJump) level = previousLevel + maxLevelJump;
    previousLevel = level;

    tocNodes.push({
      title,
      level,
      pageIndex: Math.max(0, candidate.pageIndex),
      confidence: Math.max(0.0, Math.min(1.0, candidate.totalScore)),
      source: "rule",
    });
  }

  return { tocNodes, rejectedNodes };
}

function reject(candidate: CandidateHeading, reason: string): Record<string, unknown> {
  return {
    page_index: candidate.pageIndex,
    title: candidate.text,
    reason,
    score: candidate.totalScore,
  };
}

function structuralKeyFor(title: string): string {
  const matched = STRUCTURAL_HEADING_RE.exec(title.replace(/\s+/gu, ""));
  return matched?.[1] ?? "";
}
